l1-contracts/contracts/bridge/L1Nullifier.sol

```diff
@@ -486,24 +486,29 @@ contract L1Nullifier is IL1Nullifier, ReentrancyGuard, Ownable2StepUpgradeable,
     /// @return transferData The transfer data used to finalize withdawal.
     function _verifyWithdrawal(
         FinalizeL1DepositParams calldata _finalizeWithdrawalParams
-    ) internal view returns (bytes32 assetId, bytes memory transferData) {
-        (assetId, transferData) = _parseL2WithdrawalMessage(
+    ) internal view returns (bytes32 assetId, bytes memory transferData, WithdrawalType withdrawalType) {
+        (assetId, transferData, withdrawalType) = _parseL2WithdrawalMessage(
             _finalizeWithdrawalParams.chainId,
             _finalizeWithdrawalParams.message
         );
         L2Message memory l2ToL1Message;
         {
+            address l2Sender = _finalizeWithdrawalParams.l2Sender;
             bool baseTokenWithdrawal = (assetId == BRIDGE_HUB.baseTokenAssetId(_finalizeWithdrawalParams.chainId));
-            require(
-                /// @dev for legacy function calls we hardcode the sender as the L2AssetRouter as we don't know if it is
-                /// a base token or erc20 token withdrawal beforehand,
-                /// so we have to allow that option even if we override it.
-                _finalizeWithdrawalParams.l2Sender == L2_ASSET_ROUTER_ADDR ||
-                    _finalizeWithdrawalParams.l2Sender == L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR ||
-                    _finalizeWithdrawalParams.l2Sender ==
-                    __DEPRECATED_l2BridgeAddress[_finalizeWithdrawalParams.chainId],
-                "L1N: wrong l2 sender"
-            );
+            if (withdrawalType == WithdrawalType.BaseToken) {
+                require(baseTokenWithdrawal, "L1N: token withdrawal from base token");
+                require(l2Sender == L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR, "L1N: wrong l2 sender 1");
+            } else if (withdrawalType == WithdrawalType.LegacyTokenBridgeOrSharedBridge) {
+                require(!baseTokenWithdrawal, "L1N: base token withdrawal from token bridge");
+                require(
+                    l2Sender == __DEPRECATED_l2BridgeAddress[_finalizeWithdrawalParams.chainId],
+                    "L1N: wrong l2 sender 2"
+                );
+            } else if (withdrawalType == WithdrawalType.AssetRouter) {
+                require(l2Sender == L2_ASSET_ROUTER_ADDR, "L1N: wrong l2 sender 3");
+            } else {
+                revert("L1N: wrong withdrawal type");
+            }
 
             l2ToL1Message = L2Message({
                 txNumberInBatch: _finalizeWithdrawalParams.l2TxNumberInBatch,
```

Comment:
just below this line for `l2Sender` for the message the `l2Sender` from above should be used